packages/@tailwindcss-upgrade/src/codemods/template/migrate-arbitrary-utilities.ts

```diff
@@ -0,0 +1,368 @@
+    let candidate = structuredClone(readonlyCandidate) as Writable<typeof readonlyCandidate>
+
+    // Create a basic stripped candidate without variants or important flag. We
+    // will re-add those later but they are irrelevant for what we are trying to
+    // do here (and will increase cache hits because we only have to deal with
+    // the base utility, nothing more).
+    let targetCandidate = structuredClone(candidate)
+    targetCandidate.important = false
+    targetCandidate.variants = []
+
+    let targetCandidateString = printCandidate(designSystem, targetCandidate)
+    if (baseReplacementsCache.get(designSystem).has(targetCandidateString)) {
+      let target = structuredClone(
+        baseReplacementsCache.get(designSystem).get(targetCandidateString)!,
+      )
+      // Re-add the variants and important flag from the original candidate
+      target.variants = candidate.variants
+      target.important = candidate.important
+
+      return printCandidate(designSystem, target)
+    }
+
+    // Compute the signature for the target candidate
+    let targetSignature = signatures.get(targetCandidateString)
+    if (typeof targetSignature !== 'string') continue
+
+    // Try a few options to find a suitable replacement utility
+    for (let replacementCandidate of tryReplacements(targetSignature, targetCandidate)) {
+      let replacementString = printCandidate(designSystem, replacementCandidate)
+      let replacementSignature = signatures.get(replacementString)
+      if (replacementSignature !== targetSignature) {
+        continue
+      }
+
+      // Ensure that if CSS variables were used, that they are still used
+      if (!allVariablesAreUsed(designSystem, candidate, replacementCandidate)) {
+        continue
+      }
+
+      replacementCandidate = structuredClone(replacementCandidate)
+
+      // Cache the result so we can re-use this work later
+      baseReplacementsCache.get(designSystem).set(targetCandidateString, replacementCandidate)
+
+      // Re-add the variants and important flag from the original candidate
+      replacementCandidate.variants = candidate.variants
+      replacementCandidate.important = candidate.important
+
+      // Update the candidate with the new value
+      Object.assign(candidate, replacementCandidate)
+
+      // We will re-print the candidate to get the migrated candidate out
+      return printCandidate(designSystem, candidate)
+    }
+  }
+
+  return rawCandidate
+
+  function* tryReplacements(
+    targetSignature: string,
+    candidate: Extract<Candidate, { kind: 'functional' | 'arbitrary' }>,
+  ): Generator<Candidate> {
+    // Find a corresponding utility for the same signature
+    let replacements = utilities.get(targetSignature)
+
+    // Multiple utilities can map to the same signature. Not sure how to migrate
+    // this one so let's just skip it for now.
+    //
+    // TODO: Do we just migrate to the first one?
+    if (replacements.length > 1) return
```

Comment:
as a note: I think the replacement would only be valid in this case if _all_ of the signatures matched for a given candidate.

Reply:
Yep the `replacements` array here is `signature => [class1, class2, ...]`

So this would happen in a scenario like this:

Input: 
```
bg-(--color-red-500)
```

Signature (roughly):
```
.x { background-color: oklch(63.7% 0.237 25.331) }
```

Replacements (pre computed and match the same signature):
```
['bg-red-500', 'bg-red-500/100']
```

And if you had:
```css
@utility foo { @apply bg-red-500; }
```

Then that one would also match. So right now we just skip, but we could also pick the first one, or the shortest one, or ...